### PR TITLE
Fix JsonDynamic related object serialisation problem in Jint.

### PR DESCRIPTION
**Pull request by hyzx86**
fixes: #16290

fixes: https://github.com/OrchardCMS/OrchardCore/pull/16292#issuecomment-2162818203

### CODE REVIEW

**Comment by hyzx86** [comment]
Hi @lahma , can you help review this PR? Thanks!

**Comment by lahma** [comment]
Anything related to `SetWrapObjectHandler` related to STJ sounds like Jint isn't doing good enough job, so might be worth checking if it could.

**Comment by gvkries** [comment]
> Anything related to `SetWrapObjectHandler` related to STJ sounds like Jint isn't doing good enough job, so might be worth checking if it could.

I think our problem here is more specific and it is not really a problem of Jint. We have an additional layer to support dynamics with STJ and when doing serialization inside Jint, this layer must be ignored.

I experimented with another approach by replacing the "toJSON" method, but don't know enough about Jint:

```csharp
var engine = new Engine(options =>
{
    options.Interop.MemberAccessor = (eng, target, member) =>
    {
        if (target is JsonDynamicValue || target is JsonDynamicObject || target is JsonDynamicArray)
        {
            if (member == "toJSON")
            {
                return new ClrFunction(eng, "toJSON", (thisObject, _) =>
                {
                    var wrappedInstance = ((IObjectWrapper)thisObject).Target;

                    return JConvert.SerializeObject(wrappedInstance, JOptions.Default);
                });
            }
        }

        return null;
    };
});
```

**Comment by hyzx86** [comment]
> Anything related to SetWrapObjectHandler related to STJ sounds like Jint isn't doing good enough job, so might be worth checking if it could.

I think `Jint` is good enough, it's just that STJ is too rudimentary. So we need to put a lot of effort into adapting it.

>I experimented with another approach by replacing the "toJSON" method, but don't know enough about Jint:

This would also solve the problem, my solution would be to have the `JsonDynamicXX` type converted to an STJ object, as `Jint` already handles STJ objects by default

**Comment by gvkries** [comment]
> This would also solve the problem, my solution would be to have the `JsonDynamicXX` type converted to an STJ object, as `Jint` already handles STJ objects by default

This sounds far more reasonable 👍

**Review by sebastienros — APPROVED** [review]

**Comment by hishamco** [comment]
If this rely on #16292, please don't merge until the first PR has been merged

**Comment by hyzx86** [comment]
> Suggestion, make every new introduced class sealed by default. Maybe even internal if possible.

Thanks , this problem has been solved in another PR, I just added the Jint processing to PR #16292

File: src/OrchardCore/OrchardCore.Scripting.JavaScript/JavaScriptEngine.cs
Comment (hyzx86) [inline src/OrchardCore/OrchardCore.Scripting.JavaScript/JavaScriptEngine.cs:43]: Hi @gvkries , I found during debugging that this conversion looks unnecessary, what do you think? 

image

Comment (hyzx86) [reply src/OrchardCore/OrchardCore.Scripting.JavaScript/JavaScriptEngine.cs:43]: Just to be sure, I decided to convert it to `JsonValue` anyway

Comment (gvkries) [reply src/OrchardCore/OrchardCore.Scripting.JavaScript/JavaScriptEngine.cs:43]: I've no deep knowledge of Jint, but I guess this allows to eliminate all JSON dynamic dependencies in it all together. So I think it's worth to keep it.

File: test/OrchardCore.Tests/Data/JsonDynamicTests.cs
Comment (gvkries) [inline test/OrchardCore.Tests/Data/JsonDynamicTests.cs:806]: I don't think this extra test is necessary, it does basically the same.

File: test/OrchardCore.Tests/Data/JsonDynamicTests.cs
Comment (gvkries) [inline test/OrchardCore.Tests/Data/JsonDynamicTests.cs:805]:
```suggestion
        // Assert
```

File: test/OrchardCore.Tests/Data/JsonDynamicTests.cs
Comment (hyzx86) [reply test/OrchardCore.Tests/Data/JsonDynamicTests.cs:806]: I agree. It looks like it now.

**Review by gvkries — APPROVED** [review]

File: test/OrchardCore.Tests/Data/JsonDynamicTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Data/JsonDynamicTests.cs:803]:
```suggestion
        var contentItem = GetContentTestData();
        dynamic contentExpando = new ExpandoObject();
        contentExpando.content = contentItem.Content;
        
        // Act
        var contentStr = JConvert.SerializeObject((ExpandoObject)contentExpando);
```

File: test/OrchardCore.Tests/Data/JsonDynamicTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Data/JsonDynamicTests.cs:828]:
```suggestion
        // Arrange
        using var context = new SiteContext();
        await context.InitializeAsync();
        await context.UsingTenantScopeAsync(scope =>
        {
            var getTestContent = new GlobalMethod
            {
                Name = "getTestContent",
                Method = sp => () => GetContentTestData()
            };
            var scriptingEngine = scope.ServiceProvider.GetRequiredService<IScriptingEngine>();
            var scriptingScope = scriptingEngine.CreateScope([getTestContent], scope.ServiceProvider, null, null);
            
            // Act
            var contentStr = (string)scriptingEngine.Evaluate(scriptingScope, "return JSON.stringify(getTestContent().Content)");

            // Assert
            Assert.Equal("{\"TestPart\":{\"TextFieldProp\":{\"Text\":\"test\"},\"NumericFieldProp\":{\"Value\":123},\"BooleanFieldProp\":{\"Value\":true}}}", contentStr);

            return Task.CompletedTask;
        });
```

**Review by hishamco — APPROVED** [review]

**Comment by hyzx86** [comment]
@hishamco This PR is ready to merge 😊

**Pull request merged**
